File: solutions/python/phone-number/1/phone_number.py
class PhoneNumber:
    def __init__(self, number):
        if len([k for k in number if k.isalpha() is True]) > 0: raise ValueError('letters not permitted')
        punctuation = [',','?','!',',',';',':','—','[',']','{','}','"',"'",'...']
        if len([k for k in number if k in punctuation]) > 0: raise ValueError('punctuations not permitted')

        number_formatted = [k for k in number if k.isnumeric() is True]
        number_formatted = ''.join(number_formatted)
        print(number_formatted)
        
        if len(number_formatted) < 10: raise ValueError("incorrect number of digits")
        if len(number_formatted) > 11: raise ValueError("more than 11 digits")
        if len(number_formatted) == 11:
            if number_formatted[0] == '1':
                number_formatted = number_formatted[1:]
            else:
                raise ValueError("11 digits must start with 1")
        if len(number_formatted) == 10:
            if number_formatted[0] == '0': raise ValueError('area code cannot start with zero')
            if number_formatted[0] == '1': raise ValueError('area code cannot start with one')
            if number_formatted[3] == '0': raise ValueError('exchange code cannot start with zero')
            if number_formatted[3] == '1': raise ValueError('exchange code cannot start with one')

        self.number = number_formatted
        self.area_code = number_formatted[0:3]

    def pretty(self):
        return '(' + self.number[0:3] + ')-' + self.number[3:6] + '-' + self.number[6:]

File: solutions/cpp/phone-number/1/phone_number.cpp
#include "phone_number.h"

namespace phone_number {
    phone_number::phone_number(string input) {
        if (input.substr(0, 2) == "+1") {
            number_formatted = input.substr(3, input.size());
        }
        
        for (int i = 0; i < static_cast<int>(input.size()); i++) {
            if (std::isdigit(input[i])) {
                number_ = number_ + input[i];
            }
        }

        if (number_.size() == 11 && number_[0] == '1') {
            number_ = number_.substr(1, number_.size());
        }
        
        if (number_.size() < 10 || number_.size() > 11) throw std::domain_error("Error");
        if (number_.size() == 11 && number_[0] != '1') throw std::domain_error("Error");
        if (area_code()[0] == '0' || area_code()[0] == '1') throw std::domain_error("Error");
        if (exchange_code()[0] == '0' || exchange_code()[0] == '1') throw std::domain_error("Error");
    }

    string phone_number::number() {
        return number_;
    }

    string phone_number::area_code() {
        return number_.substr(0, 3);
    }

    string phone_number::exchange_code() {
        return number_.substr(3, 6);
    }

    phone_number::operator string() const {
        return number_formatted;
    }
}  // namespace phone_number
